### ISSUE TITLE
Precompile fails with >64 threads

### ISSUE THREAD

**Issue by jmulcahy** [JuliaSIMD/CheapThreads.jl#22]
This line fails: https://github.com/JuliaSIMD/CheapThreads.jl/blob/master/src/CheapThreads.jl#L21 if `dynamic_thread_count()` is greater than 64. For example, `(1 << (65 - 1)) - 1 == -1`, which can't be stuffed into a `UInt128`. Casting the leftmost 1 to a `UInt128` at least allow the precompile to work, but I don't know if that is actually correct in the context of the package.

**Comment by jmulcahy** [comment]
Can `reset_workers!` and the call to it in `__init__` just be removed? It isn't used anywhere from what I can tell.

Is `WORKERS` meant to be a bitfield describing the status of each thread? If so, even a `UInt128` won't work when there are >128 threads. `BigInt` would work, but is slow even with bit bashing tricks.

This is currently preventing me from using LoopVectorization on certain machines.

**Comment by chriselrod** [comment]
Yes, the integer is a bitfield describing the status of each thread.

Regarding supporting >128 threads, it seems like I'll need to add `UInt256` and `UInt512` support, or maybe cap the number of threads used to 128 as a temporary workaround (better than throwing an error).
The actual integer type used will depend on how many threads the host computer has, so adding `UInt512` (for example) support won't impact performance on systems with fewer threads.

**Comment by jmulcahy** [comment]
Oops, I see, `WORKERS` is indirectly referenced via `worker_pointer`. Maybe `WORKERS` could be an array or tuple of `UInt128`s? It seems like a tuple of several could be still be somewhat efficient with compiler unrolling. Definitely more cumbersome, though.

Maybe a stopgap is to just cap the number of threads to 64/128 for now?

**Comment by chriselrod** [comment]
Yeah, actually a tuple of `UInt128` would be best. Most computers support 128 bit atomic instructions, and we don't really need more than that to be atomic at a time (i.e., 2x 128 bit atomic operations is fine).

**Comment by chriselrod** [comment]
For now, this should fix <=128 threads:
https://github.com/JuliaSIMD/CheapThreads.jl/commit/efd59078593373c71938f5699daeb5713f248323

**Comment by jmulcahy** [comment]
Awesome, thanks.

**Comment by jmulcahy** [comment]
I notice with this change that:
```julia
julia> (one(UInt128) << (128 - one(UInt128))) - one(UInt128)
0x7fffffffffffffffffffffffffffffff

julia> (one(UInt128) << (256 - one(UInt128))) - one(UInt128)
0xffffffffffffffffffffffffffffffff
```
So this compiles with >128 threads, but it seems like it won't work right. Does a 128 in the `min` in `dynamic_thread_count` fix things?

Also, is the leading bit meant to be 0 in the 128 case? The mismatch between `dynamic_thread_count` and the number of threads managed by `WORKERS` seems odd, but I could see that being on purpose so that you're leaving at least one thread open for the caller.

**Comment by chriselrod** [comment]
> Also, is the leading bit meant to be 0 in the 128 case? The mismatch between `dynamic_thread_count` and the number of threads managed by `WORKERS` seems odd, but I could see that being on purpose so that you're leaving at least one thread open for the caller.

Yes. It's strongly recomended that when using `ThreadingUtilities` and `CheapThreads` that, when wanting to run a program on `N` threads, you have the main thread launch `N-1` threads, and then have it run the `N`th chunk itself.
`LoopVectorization.@avxt`, `Octavian.matmul!`, and `CheapThreads.batch`/`CheapThreads.@batch` all do this.

The reason for this is that `ThreadingUtilities.wait` uses a spin and wait loop. That thread is blocked but does not go to sleep.
It's thus best to minimize the amount of time spent having any thread `wait`ing, e.g. by having it do one of the chunks of worse first.

> No idea what the implications are.

That should be fine.

**Comment by jmulcahy** [comment]
Ok, thanks for the explanation. The big picture makes a lot of sense.

I'll test this tomorrow with my Tullio / LoopVectorization code on a 128 core / 256 thread machine.

**Comment by chriselrod** [comment]
Given that 256 threads didn't work locally (on a 10 core/20 thread machine), I'm not optimistic.
I'll leave this open until `CheapThreads` supports >128 threads by either switching to
1. Using `NTuple{N,UInt128}` to describe worker states.
2. Switches to decentralized thread management.

Although I'd probably create a new library if/when I wanted to pursue "2.".

Also, I'm curious how well threading works generally on your 128 / 256 thread machine.
A lot of code I work on that uses multithreading (built on `Threads.@spawn` or `Threads.@threads`) ends up with 50-80% GC time on an 18 core/36 thread machine, and I imagine this will only get worse.

**Comment by jmulcahy** [comment]
Yeah, the GC can rapidly become an issue in heavily allocating code. I've generally pursued threading over `Distributed` due to long process start up times and/or needing reductions over many fairly large arrays. In very highly parallel code I've mostly opted for higher levels of abstraction than direct `Threads.@spawn` calls. For the problems it fits into, `Tullio` makes scaling pretty easy. `ThreadsX` and `Transducers` require some more effort to avoid allocations, but I've used them to solve some problems that would be impossible in Matlab/Python and difficult in C/C++.

I have noticed some weird plateaus that I think are hardware specific. The machine is a dual socket AMD Epyc server, and things start to seem to scale worse after 64 threads. I'm guessing it has something to do with spilling threads across sockets, but even threaded code with almost no inter-tread communication slows down. It could also be something with AMD's chiplet architecture, I suppose. I have no idea how Linux assigns threads to cores with this somewhat unusual hardware set up.

It would be nice if Julia's GC scaled better with multithreaded code, but I'm sure that's a huge undertaking. Some libraries that helped with setting up preallocation would help. For example, I think thread-safe lazy memory pools would be really nice for avoiding excessive preallocation in code with nested tasks.

**Comment by jmulcahy** [comment]
The Tullio / LoopVectorization code ran, although I didn't get a chance to check for correctness.
